lib/librte_regexdev/rte_regexdev_core.h

```diff
@@ -0,0 +1,168 @@
+/**< @internal Trigger regex self test. */
+
+typedef int (*regex_dev_dump_t)(struct rte_regex_dev *dev, FILE *f);
+/**< @internal Dump internal information about the regex device. */
+
+typedef int (*regex_dev_enqueue_t)(struct rte_regex_dev *dev, uint16_t qp_id,
+				   struct rte_regex_ops **ops, uint16_t nb_ops);
+/**< @internal Enqueue a burst of scan requests to a queue on regex device. */
+
+typedef int (*regex_dev_dequeue_t)(struct rte_regex_dev *dev, uint16_t qp_id,
+				   struct rte_regex_ops **ops,
+				   uint16_t nb_ops);
+/**< @internal Dequeue a burst of scan response from a queue on regex device. */
+
+/**
+ * regex device operations
+ */
+struct rte_regex_dev_ops {
+	regex_dev_info_get_t dev_info_get;
+	regex_dev_configure_t dev_configure;
+	regex_dev_qp_setup_t dev_qp_setup;
+	regex_dev_start_t dev_start;
+	regex_dev_stop_t dev_stop;
+	regex_dev_close_t dev_close;
+	regex_dev_rule_db_update_t dev_rule_db_update;
+	regex_dev_rule_db_import_t dev_db_import;
+	regex_dev_rule_db_export_t dev_db_export;
+	regex_dev_xstats_names_get_t dev_xstats_names_get;
+	regex_dev_xstats_get_t dev_xstats_get;
+	regex_dev_xstats_by_name_get_t dev_xstats_by_name_get;
+	regex_dev_xstats_reset_t dev_xstats_reset;
+	regex_dev_selftest_t dev_selftest;
+	regex_dev_dump_t dev_dump;
+
+	/** Reserved for future extension */
+	void *reserved[5];
+};
+
+/**
+ * @internal
+ * The data part, with no function pointers, associated with each regex device.
+ *
+ * This structure is safe to place in shared memory to be common among different
+ * processes in a multi-process configuration.
+ */
+struct rte_regex_dev_data {
+	char name[RTE_REGEX_NAME_MAX_LEN]; /**< Unique identifier name */
+	uint16_t nb_qp; /**< Number of queue pairs. */
+	void *dev_private; /**< PMD-specific private data. */
+	struct rte_regex_dev_config dev_conf;
+	/**< Configuration applied to device. */
+	uint16_t port_id;           /**< Device [external] port identifier. */
+	uint64_t reserved_64s[4]; /**< Reserved for future fields */
+	void *reserved_ptrs[4];   /**< Reserved for future fields */
+} __rte_cache_aligned;
+
+/**
+ * @internal
+ * The generic data structure associated with each RegEx device.
+ *
+ * Pointers to burst-oriented packet receive and transmit functions are
+ * located at the beginning of the structure, along with the pointer to
+ * where all the data elements for the particular device are stored in shared
+ * memory. This split allows the function pointer and driver data to be per-
+ * process, while the actual configuration data for the device is shared.
+ */
+struct rte_regex_dev {
+	regex_dev_enqueue_t enqueue;
+	regex_dev_dequeue_t dequeue;
+	struct rte_regex_dev_data *data;  /**< Pointer to device data. */
+	const struct regex_dev_ops *dev_ops; /**< Functions exported by PMD */
+	struct rte_device *device; /**< Backing device */
+	uint64_t reserved_64s[4]; /**< Reserved for future fields */
+	void *reserved_ptrs[4];   /**< Reserved for future fields */
```

Comment:
@orikam 
not sure why future fields should be of two different types as uint64_t and void* to be defined now?

__rte_cache_aligned is already defined, so reserved fields should not be added.

Reply:
we can remove the reserve due to the fact that it is internal structure.
but basically since we don't support ABI breakage only once a year. the cache_aligned doesn't necessarily gives us enough extra space.

Reply:
@orikam 
how one developer tracks dpdk release version done with reserved array size, commit history all to make sure.
This is really very fragile way to ensure ABI compatibility.
Giving little more thought _data is internal structure of the library core for its internal book keeping, it doesn't need ABI compatibility.